test/compare_with_wikidata_test.rb

```diff
@@ -1,7 +1,28 @@
 require 'test_helper'
 
 describe 'CompareWithWikidata' do
+  subject { CompareWithWikidata::DiffOutputGenerator.new(
+    mediawiki_site: 'www.wikidata.org',
+    page_title: 'FooBar',
+  ) }
+
```

Comment:
Having to add a new `subject` to an existing test is a bit of a red flag to me. If we want to add a test like this I think I'd add it either to a new test file entirely, or at least to a new `describe` block.

*However*, I'm also not entirely convinced that we do want tests like this. They're very implementation dependent (e.g. if we were to replace the `CSV.parse(fix_encoding(csv_text))` with `CSV.parse(csv_text.detect_encoding!)`, these tests would fail), and don't really exercise the actual problem we were facing (i.e. a comparison blowing up). 

I'd be a lot more comfortable if we added a test that actually takes two datasets with different encoding, and showed the comparison succeeding rather than exploding.

Reply:
As background, this wasn't how I really wanted to do this fix: it would be nice to do the encoding fixing in `daff_diff`, so that however that function is used in future can benefit from the change. And it's much easier to write specific tests for whether the comparison works or not. However, by that point any strings are spread across different cells - you could do something like:

```
    daff_table.map do |row|
      row.map do |cell|
        next cell unless cell.instance_of? String
        next cell unless cell.encoding == Encoding::ASCII_8BIT
        cell.dup.detect_encoding!
      end
    end
```

(full example [here](https://github.com/everypolitician/compare_with_wikidata/compare/fix-unknown-character-encoding2?expand=1)) but when you're considering a cell at a time, the character guessing of charlock_holmes works surprisingly badly (e.g. detecting UTF-8 as ISO-8859-1 in cases you're have thought were obvious); really you want to do the character detection on all the strings in the dataset to help its guesses. I guess you could still do it within `daff_diff`, by extracting all the strings in a dataset, concatenating them and guessing the encoding from that, but that seems pretty hacky.

However, it sounds as if you're OK with the implementation in this PR, and just asking for the tests to be more helpful. This wasn't particularly obvious because of how `CompareWithWikidata.run! ` is written, but I've added a refactoring commit first it a bit to make that possible. Also, I realised that there were lots of existing tests for `csv_from_url` so I've added a couple there too.